### PR TITLE
ARROW-16507: [CI][C++] Use system gtest with mamba/conda 

### PR DESCRIPTION
**Pull request by assignUser**
The gtest package for Windows provided by conda-forge doesn't provide `GTestConfig.cmake`.
See also: https://github.com/conda-forge/gtest-feedstock/blob/main/recipe/bld.bat

And `FindGTest.cmake` provided by CMake can't find `gtest_dll.dll` that is a shared
library version of GoogleTest.
See also: https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindGTest.cmake

It means that we can find only static version of GoogleTest on Windows with Conda
without a custom `FindGTestAlt.cmake`.

Shared library version `arrow_flight_testing` requires shared library version GoogleTest on
Windows because it defines `arrow::flight::FlightTest` that inherits `testing::Test`.
See also: https://github.com/apache/arrow/blob/master/cpp/src/arrow/flight/test_definitions.h

We must use the same library type for them on Windows.

To support `ARROW_FLIGHT=ON`/`ARROW_BUILD_SHARED=ON`/`ARROW_BUILD_STATIC=OFF`/
`ARROW_BUILD_TESTS=ON` with static version of GoogleTest, we need to build a static library not
shared library for  `arrow_flight_testing`.

### CODE REVIEW

**Comment by kou** [comment]
It seems that the gtest package for Windows doesn't include CMake package files:

* https://anaconda.org/conda-forge/gtest
* https://github.com/conda-forge/gtest-feedstock/blob/main/recipe/bld.bat

It seems that `FindGTest.cmake` https://cmake.org/cmake/help/latest/module/FindGTest.html bundled in CMake 3.23 or later fits our use case:

* It uses `GTestConfig.cmake` provided by Google Test if possible
* It defines not only `GTest::gtest` and `GTets::gtest_main` but also `GTest::gmock`

How about trying it?

```diff
diff --git a/cpp/cmake_modules/ThirdpartyToolchain.cmake b/cpp/cmake_modules/ThirdpartyToolchain.cmake
index 865406e705..7dfe388bb7 100644
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1969,11 +1969,16 @@ macro(build_gtest)
 endmacro()
 
 if(ARROW_TESTING)
+  if(CMAKE_VERSION VERSION_LESS 3.23)
+    set(GTEST_USE_CONFIG FALSE)
+  else()
+    set(GTEST_USE_CONFIG TRUE)
+  endif()
   resolve_dependency(GTest
                      REQUIRED_VERSION
                      1.10.0
                      USE_CONFIG
-                     TRUE)
+                     ${GTEST_USE_CONFIG})
 
   if(NOT GTEST_VENDORED)
     # TODO(wesm): This logic does not work correctly with the MSVC static libraries
```

**Comment by assignUser** [comment]
With cmake 3.23.1 this does not work on my machine, setting `USE_CONFIG` to `FALSE` manually does work though.

**Comment by kou** [comment]
Oh, sorry. `GTEST_USE_CONFIG` value was inverted:

```diff
--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -1969,11 +1969,16 @@ macro(build_gtest)
 endmacro()
 
 if(ARROW_TESTING)
+  if(CMAKE_VERSION VERSION_LESS 3.23)
+    set(GTEST_USE_CONFIG TRUE)
+  else()
+    set(GTEST_USE_CONFIG FALSE)
+  endif()
   resolve_dependency(GTest
                      REQUIRED_VERSION
                      1.10.0
                      USE_CONFIG
-                     TRUE)
+                     ${GTEST_USE_CONFIG})
 
   if(NOT GTEST_VENDORED)
     # TODO(wesm): This logic does not work correctly with the MSVC static libraries
```

**Comment by assignUser** [comment]
Thanks for the quick response, that worked locally! I'll check with some crossbow builds.

**Comment by assignUser** [comment]
@github-actions crossbow submit verify-rc-source-windows

**Comment by assignUser** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by assignUser** [comment]
@github-actions crossbow submit conda-linux-gcc-py37-cpu-r40

**Comment by assignUser** [comment]
@github-actions crossbow submit conda-osx-clang-py37-r40

**Comment by assignUser** [comment]
@github-actions crossbow submit test-build-vcpkg-win

**Comment by pitrou** [comment]
It seems AppVeyor is still broken with this PR? cc @kou

**Comment by assignUser** [comment]
In the Appveyor build cmake 3.17 is used, so `GTEST_USE_CONFIG` is set to `TRUE`. And as the conda package still has no Config.cmake the error occures. But as I understand the docs the [`FindGTest.cmake`](https://cmake.org/cmake/help/v3.0/module/FindGTest.html) has been available since cmake 3.0 so we could either set it to always be `FALSE` or not set it explicitly at all in which case it [would fallback](https://cmake.org/cmake/help/v3.17/command/find_package.html) to looing for `GTestConfig.cmake` if it was not able to locate Gtest via the bundled FindGtest. But I do not know if this might have unwanted repercussions on other builds. @kou ?

**Comment by kou** [comment]
How about just stopping pinning CMake?

```diff
diff --git a/ci/appveyor-cpp-setup.bat b/ci/appveyor-cpp-setup.bat
index 936306296f..bbe2c38d1e 100644
--- a/ci/appveyor-cpp-setup.bat
+++ b/ci/appveyor-cpp-setup.bat
@@ -71,7 +71,6 @@ if "%JOB%" NEQ "Build_Debug" (
   mamba create -n arrow -q -y -c conda-forge ^
     --file=ci\conda_env_python.txt ^
     %CONDA_PACKAGES%  ^
-    "cmake=3.17" ^
     "ninja" ^
     "nomkl" ^
     "pandas" ^
```

This was introduced by https://github.com/apache/arrow/pull/7865 / ARROW-9599 as a workaround for CMake 3.18.

The problem may be resolved with the latest CMake.

**Comment by assignUser** [comment]
Thank you for your work on this!

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by xhochy** [comment]
Shouldn't we fix the conda package instead? This sounds like a bug in it.

**Comment by kou** [comment]
In general, it's better that the Conda package for Windows provides CMake's package file. But I think that it's difficult because https://github.com/conda-forge/gtest-feedstock/blob/main/recipe/bld.bat mixes multiple CMake configurations.

I think that a real fix for this is stopping `class ARROW_FLIGHT_EXPORT arrow::flight::FlightTest : public testing::Test` in https://github.com/apache/arrow/blob/master/cpp/src/arrow/flight/test_definitions.h that is introduced by #12499/ARROW-15707. (Cc: @lidavidm)
If we implement shared tests by header only library, macro or something instead of shared library, we can use static library version of GoogleTest for `arrow_flight_testing.dll`.

Anyway, I want to merge this workaround as soon as possible because this breaks our CI since 2022-05-07. (I should have checked carefully before I merge #13083/ARROW-16490.)

**Comment by kou** [comment]
@github-actions crossbow submit test-debian-*-cpp-* test-fedora-35-cpp test-ubuntu-20.04-cpp-bundled test-ubuntu-22.04-cpp verify-rc-source-cpp-linux-almalinux-8-amd64 verify-rc-source-cpp-linux-ubuntu-* verify-rc-source-integration-linux-ubuntu-* verify-rc-source-integration-macos-conda-amd64

**Comment by kou** [comment]
@github-actions crossbow submit -g nightly-tests -g nightly-packaging -g nightly-release

**Comment by pitrou** [comment]
Can you trigger an AppVeyor build on this PR?

**Comment by lidavidm** [comment]
> I think that a real fix for this is stopping `class ARROW_FLIGHT_EXPORT arrow::flight::FlightTest : public testing::Test` in https://github.com/apache/arrow/blob/master/cpp/src/arrow/flight/test_definitions.h that is introduced by #12499/[ARROW-15707](https://issues.apache.org/jira/browse/ARROW-15707). (Cc: @lidavidm)
> If we implement shared tests by header only library, macro or something instead of shared library, we can use static library version of GoogleTest for `arrow_flight_testing.dll`.

Hmm, part of the motivation was to try to save some compile time instead of having to include the full definition of all tests. But if it's causing problems I can move it into a header only library instead.

**Comment by pitrou** [comment]
Perhaps it can simply be a helper class that doesn't inherit from `testing::Test` instead? (I assume the problem is different DLL exports for base and derived classes?)

**Comment by lidavidm** [comment]
Ah, that could work. We still need to link the testing library to gtest but at least we can avoid inheritance issues.

**Comment by kou** [comment]
This is ready to merge.

AppVeyor build: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/43563294

`arrow-flight-sql-test` is sometimes stuck on macOS. e.g.:  https://github.com/ursacomputing/crossbow/runs/6447575030?check_suite_focus=true#step:7:6883
But this is not related to this change because it's caused without this change. e.g.: [verify-rc-source-python-macos-conda-amd64 nightly at 2022-05-09](https://github.com/ursacomputing/crossbow/runs/6353247542?check_suite_focus=true)

Note that the nightly job at 2022-05-09 was canceled by GitHub Actions' default 6 hours timeout because we ran tests without ctest's `--timeout`. The `arrow-flight-sql-test` failure in this pull request is stopped by `ctest` because I added `--timeout` in this pull request: https://github.com/apache/arrow/pull/13101/files#diff-82f32771b6ac9daaadbfb77bd77062d63e131ad18ecd9014ff90eff17865964fR638

File: cpp/src/arrow/flight/CMakeLists.txt [hunk 1]
```diff
@@ -25,27 +25,52 @@ if(WIN32)
...
 endif()
 if(NOT ARROW_GRPC_USE_SHARED)
   message(STATUS "Linking Arrow Flight tests statically due to static gRPC")
-  set(ARROW_FLIGHT_TEST_LINKAGE
-      "static"
-      PARENT_SCOPE)
+  set(ARROW_FLIGHT_TEST_LINKAGE "static")
+endif()
+if(ARROW_TESTING)
+  if(WIN32)
```
Comment (pitrou) [inline cpp/src/arrow/flight/CMakeLists.txt:38]: Hopefully this can be removed next in @lidavidm 's PR.

**Comment by pitrou** [comment]
Oops, should have fixed the PR title before merging. Too bad.

**Pull request closed without merging**